Lib/typing.py

```diff
@@ -1060,15 +1060,27 @@ def _evaluate(self, globalns, localns, type_params=_sentinel, *, recursive_guard
                 globalns = getattr(
                     sys.modules.get(self.__forward_module__, None), '__dict__', globalns
                 )
+
+            # type parameters require some special handling,
+            # as they exist in their own scope
+            # but `eval()` does not have a dedicated parameter for that scope.
+            # For classes, names in type parameter scopes should override
+            # names in the global scope (which here are called `localns`!),
+            # but should in turn be overridden by names in the class scope
+            # (which here are called `globalns`!)
             if type_params:
-                # "Inject" type parameters into the local namespace
-                # (unless they are shadowed by assignments *in* the local namespace),
-                # as a way of emulating annotation scopes when calling `eval()`
-                locals_to_pass = {param.__name__: param for param in type_params} | localns
-            else:
-                locals_to_pass = localns
+                if self.__forward_is_class__:
+                    globalns, localns = dict(globalns), dict(localns)
+                    for param in type_params:
+                        param_name = param.__name__
+                        if param_name not in globalns:
```

Comment:
If you want to gain one line (though I'm not sure you're gaining in clarity, depends on the reader I'd say):

```suggestion
                        if (param_name := param.__name__) not in globalns:
```

Reply:
Thanks for the suggestion, but I think I'll pass on this :-)